[Wang Ruofan] iP

Ruofan1023 opened the pull request:
### DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
 * text-based
 * easy to learn
 * ~~FAST~~ *SUPER FAST* to use
 
All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!
Features:
 - [x] Managing tasks
 - [x] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Ruofan1023 committed:
level 1

Ruofan1023 committed:
level two

Ruofan1023 committed:
level two doc

Ruofan1023 committed:
Level-3

Ruofan1023 committed:
Level 3 doc

Ruofan1023 committed:
Level 4

Ruofan1023 committed:
small grammatical error fix

Ruofan1023 committed:
Level 4 UI improvement

Ruofan1023 committed:
bug fix

Ruofan1023 committed:
text ui test

Ruofan1023 committed:
level 5

Ruofan1023 committed:
level 6

Ruofan1023 committed:
enum version

Ruofan1023 committed:
updated javadoc

Ruofan1023 committed:
Add save feature

Duke saves the tasks in {root directory}/data/duke.txt, and loads the tasks from disk upon init.

Ruofan1023 committed:
Update davadoc

Ruofan1023 committed:
Format code

Ruofan1023 committed:
Enable duke to understand dates

Ruofan1023 committed:
Merge branch 'branch-Level-7'

Ruofan1023 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Main.java
#	src/main/java/Task.java

Ruofan1023 committed:
Complete A-MoreOOP

Ruofan1023 committed:
Complete A-Packages

Ruofan1023 committed:
Add JUnit Tests

Ruofan1023 committed:
Save changes

Ruofan1023 committed:
Add JavaDoc comments

Ruofan1023 committed:
Tweak the code to comply with a coding standard

Ruofan1023 committed:
Give users a way to find a task by searching for a keyword

Ruofan1023 committed:
Merge branch 'branch-A-JavaDoc'

Ruofan1023 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Task.java
#	src/test/java/duke/DukeTest.java

Ruofan1023 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Task.java

Ruofan1023 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Ruofan1023 committed:
Complete A-Gradle

Ruofan1023 committed:
Complete A-CheckStyle

Ruofan1023 committed:
Merge branch 'branch-A-CheckStyle'

Ruofan1023 committed:
Complete Level-10

Ruofan1023 committed:
Merge branch 'branch-Level-10'

Ruofan1023 committed:
Use varargs to search for keywords

Ruofan1023 committed:
Merge branch 'branch-A-Varargs'

boonhaii reviewed (commented):
Overall I thought the code was clean and there weren't any major Coding Standard violations found. I think it should be good to merge!

Ruofan1023 commented:
> Overall I thought the code was clean and there weren't any major Coding Standard violations found. I think it should be good to merge!

Thanks for the review!

Harry-Gao-H reviewed (commented):
I like your code, and I think it follows coding standards well 👍  LGTM

kaixin-hc commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,146 @@
+            throw new UnrecognizableCommandException();
+        }
+        try {
+            String[] lst = task.split(" /by ");
```
Code LGTM! Nice job :) I was wondering about this specific section though -- would an input of `event /by 1999-10-30` work even though it doesn't have a description? I don't see the EmptyDescriptionException

Ruofan1023 committed:
Add Launcher.java. Update build.gradle

Ruofan1023 committed:
Complete A-Assertions

Add important assumption checks

Add mostly null-pointer/empty string checks since these functions assume
that the input is not null or empty and passing such parameters in will
otherwise trigger errors.

Ruofan1023 committed:
Merge branch 'branch-A-Assertions'

Ruofan1023 committed:
Complete A-CodeQuality

Critically examine the code and refactor to improve the code quality
where necessary.

Increase code readability by avoiding deep nesting, complicated
expressions, magic numbers, and by using the SLAP and the KISSing
principle, etc.

Ruofan1023 committed:
Update A-Assertions

Ruofan1023 committed:
Complete A-Streams

Looping through tasks to find result is wordy and slightly complicated
to implement.

Use stream to filter matched tasks achieves the same purpose with an
easy implementation and increases readability of the code.

Ruofan1023 committed:
Merge pull request #1 from Ruofan1023/branch-A-Assertions

Update A-Assertions

Ruofan1023 committed:
Merge branch 'master' of https://github.com/Ruofan1023/ip

* 'master' of https://github.com/Ruofan1023/ip:
  Update A-Assertions

Ruofan1023 committed:
Merge branch 'master' into branch-CodeQuality

Ruofan1023 committed:
Merge pull request #2 from Ruofan1023/branch-CodeQuality

Complete A-CodeQuality

Ruofan1023 committed:
Merge branch 'master' of https://github.com/Ruofan1023/ip

* 'master' of https://github.com/Ruofan1023/ip:
  Complete A-CodeQuality

Ruofan1023 committed:
Merge branch 'master' into branch-A-Streams

Ruofan1023 committed:
Merge pull request #3 from Ruofan1023/branch-A-Streams

Complete A-Streams

Ruofan1023 committed:
Fix issue with function naming mismatch

Ruofan1023 committed:
Complete BCD-Extension

Add C-Help to show a help page with command "help".

Ruofan1023 committed:
Merge branch 'branch-C-Help'

Ruofan1023 committed:
Format style to pass checkstyle

CommanderW324 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,122 @@
+     * @return Deleted task.
+     */
+    public String showDeletedTask(Task toBeDeleted) {
+        assert (toBeDeleted != null);
```
I like with how you use assert to simplify the code, but is it safe to use it for this public method ?

Samuel-bit-prog commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Throws when the command is not recognised by duke.
+ */
+class UnrecognizableCommandException extends DukeException {
```
Maybe can use enum for all the child class as a suggestion. Overall the code looks very good. Could not find any coding violations.

Samuel-bit-prog reviewed (commented):
The code is very clean and esay to understand. keep up the good work!

CommanderW324 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,152 @@
+     * @param keywords Keywords to be found among all tasks.
+     * @return A list of matched tasks.
+     */
+    public List<Task> findMatchedTasks(String... keywords) {
+        Stream<Task> matchedTasksStream = Stream.empty();
+        for (String keyword : keywords) {
+            if (keyword.isEmpty()) {
+                continue;
+            }
+            matchedTasksStream = this.tasks.stream()
+                    .filter((task) -> task.getDescription().contains(keyword));
+        }
+        return matchedTasksStream.collect(Collectors.toList());
+    }
+}
```
I like how you use the stream idea to search, leading to a very concise code and readable !

CommanderW324 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    public void saveProgress() {
+        try {
+            FileWriter fw = new FileWriter("data/duke.txt");
+            for (Task task : tasks) {
+                fw.write(task + "\n");
+            }
+            fw.close();
+        } catch (IOException e) {
+            System.out.println("No File Found");
+            System.exit(-1);
+        }
```
Shouldn't adding a finally block to close the Filewriter instead of closing it inside the try block better ? (In the case of an exception occuring in the try code).

CommanderW324 reviewed (commented):
In my opinion, you did a very good job on having concise and readable code !

Ruofan1023 committed:
Complete A-BetterGui

Ruofan1023 committed:
Complete A-BetterGui

Ruofan1023 committed:
Complete A-UserGuide

Ruofan1023 committed:
Update product website
